File: src/DashSpec.Core/Parsing/LayoutModuleScopeValidator.cs
using DashSpec.Core.Model;

namespace DashSpec.Core.Parsing;

internal static class LayoutModuleScopeValidator
{
    public static void EnsureMatchesIncludeSite(
        LayoutBoardDefinition board,
        LayoutScope expected,
        string context)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(context);

        if (board.ModuleScope is null)
        {
            return;
        }

        if (board.ModuleScope != expected)
        {
            throw new DashSpecParseException(
                $"{context}: layout module declares scope {FormatScope(board.ModuleScope.Value)} " +
                $"but was included for {FormatScope(expected)}.");
        }
    }

    private static string FormatScope(LayoutScope scope) =>
        scope switch
        {
            LayoutScope.Toolbar => "toolbar",
            LayoutScope.Tab => "tab",
            LayoutScope.Card => "card",
            _ => scope.ToString(),
        };
}

File: src/DashSpec.Modeling.Parse/Layout/LayoutModuleScopeValidator.fs
namespace DashSpec.Modeling.Parse.Layout

open System
open DashSpec.Modeling.Core

module LayoutModuleScopeValidator =

    let private formatScope scope =
        match scope with
        | LayoutScope.Toolbar -> "toolbar"
        | LayoutScope.Tab -> "tab"
        | LayoutScope.Page -> "page"
        | LayoutScope.Card -> "card"

    let ensureMatchesIncludeSite (board: LayoutBoardDefinition) (expected: LayoutScope) (context: string) =
        if String.IsNullOrWhiteSpace context then
            invalidArg "context" "Context is required."

        match board.ModuleScope with
        | None -> ()
        | Some scope when scope <> expected ->
            raise (DashSpecParseException($"{context}: layout module declares scope {formatScope scope} but was included for {formatScope expected}."))
        | Some _ -> ()
